### PR TITLE
Iox #415 refactor service registry

### PR DESCRIPTION
**Pull request by MatthiasKillat**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CHANGELOG.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Reviewer notes

Only change the internal implementation of `ServiceRegistry` to use a vector, but NOT the interface (except for extension with one method `removeAll`).

## References

- Closes nothing

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 1]
```diff
@@ -38,25 +39,42 @@ class ServiceRegistry
...
-    /// @brief Removes given service description from registry if service is found
+    /// @brief Removes given service description from registry if service is found,
+    ///        in case of multiple occurrences only one occurrence is removed
     /// @param[in] serviceDescription, service to be removed
     void remove(const capro::ServiceDescription& serviceDescription) noexcept;
 
+    /// @brief Removes given service description from registry if service is found,
+    ///        all occurences are removed
+    /// @param[in] serviceDescription, service to be removed
+    void removeAll(const capro::ServiceDescription& serviceDescription) noexcept;
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:76]: We may also want a method to remove all services (i.e. clear it). But this is part of the next step: improving the interface.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 2]
```diff
@@ -38,25 +39,42 @@ class ServiceRegistry
...
+        }
+
         capro::ServiceDescription serviceDescription{};
-        ReferenceCounter_t referenceCounter = 0U;
+        ReferenceCounter_t count{0U};
     };
 
     /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
-    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
+    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 5000U;
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:56]: Deciding how to configure the values is also part of an interface change PR.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 3]
```diff
@@ -17,87 +17,83 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/roudi/service_registry.cpp:18]: Remark for reviewers:
The diff is almost useless, all implementations where changed, i.e. nothing except the interfaces remain.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 4]
```diff
@@ -445,4 +465,109 @@ TEST_F(ServiceRegistry_test, AddingVariousServiceDescriptionAndGetServicesDoesNo
...
+    {
+        auto id = randomString();
+        ServiceDescription sd(fixedId, fixedId, id);
+
+        auto result = sut.add(sd);
+        if (result.has_error())
+        {
+            break;
+        }
+        lastAdded = sd;
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:543]: Needs to be stored for removal afterwards since we do not know when the registry is full if we add random values (duplicates occupy no extra space).

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 5]
```diff
@@ -38,25 +39,40 @@ class ServiceRegistry
     };
 
     using ReferenceCounter_t = uint64_t;
+
     struct ServiceDescriptionEntry
     {
+        ServiceDescriptionEntry(const capro::ServiceDescription& sd, ReferenceCounter_t count)
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:45]: Please move all implementations into cpp file.

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 6]
```diff
@@ -17,87 +17,83 @@
...
+        auto& entry = m_serviceDescriptions[index];
+        entry->count++;
+        return cxx::success<>();
+    }
+
+    // entry does not exist, find a free slot if it exists
+
+    // fast path to a free slot (which was occupied by previously removed entry),
+    // prefer to fill entries close to the front
+    if (m_freeIndex != NO_INDEX)
```
Comment (elfenpiff) [inline iceoryx_posh/source/roudi/service_registry.cpp:39]: Actually one could improve this even further.
1. Introduce the fifo which can be found in `internal/concurrent/fifo.hpp` (its single push/single pop with an atomic) as `concurrent::FiFo<uint32_t, MAX_SERVICE_DESCRIPTIONS> m_freeIndices`
2. move all indices which are freed in there with just `m_freeIndices.push(indexToFree)`.
3. At this position we take a potential free index with `m_freeIndices.pop` and if it is not a nullopt use this one.

The solution above would not even increase the lines of code or would make the code more complex but would optimize it even further.

My personal opinion either keep it simple and make it work first before optimizing it but when one goes that route than take the low hanging fruits with you.

Comment (elfenpiff) [reply iceoryx_posh/source/roudi/service_registry.cpp:39]: Or we could use the index queue ... wouldn't than this service registry be thread safe in a lock free manner?! I have not really thought this through

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 5 above)
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:45]: It looks like you are always using `1` as count for the `ServiceDescriptionEntry` wouldn't it make more sense to set it always to one and remove the second argument?
If we need it later we always can add it again - but YAGNI.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 7]
```diff
@@ -38,25 +39,40 @@ class ServiceRegistry
...
+        }
+
         capro::ServiceDescription serviceDescription{};
-        ReferenceCounter_t referenceCounter = 0U;
+        ReferenceCounter_t count{0U};
     };
 
     /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
+    ///            increase limit to at least >= 1000
     static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:57]: Why not set it here directly to `iox::MAX_NUMBER_OF_SERVICES` otherwise I may fear that you brake master.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 8]
```diff
@@ -72,11 +88,61 @@ class ServiceRegistry
...
-    ::std::multimap<capro::IdString_t, uint64_t> m_instanceMap;
-    ::std::multimap<capro::IdString_t, uint64_t> m_eventMap;
-    ServiceDescriptionVector_t m_serviceDescriptionVector;
+    using Entry_t = cxx::optional<ServiceDescriptionEntry>;
+    using ServiceDescriptionContainer_t = cxx::vector<Entry_t, MAX_SERVICE_DESCRIPTIONS>;
+
+    static constexpr uint32_t NO_INDEX = MAX_SERVICE_DESCRIPTIONS;
+
+    // tag type for internal overloads
+    struct Wildcard
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:97]: Why not use the definition from: `iceoryx_posh/include/iceoryx_posh/capro/service_description.hpp:33`?

This is compile time and has no overhead.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 9]
```diff
@@ -17,87 +17,83 @@
...
-            }
-            else if (it->second > index)
-            {
-                // update index due to removed element
-                it->second--;
-            }
-            it++;
+            entry.reset();
+            // reuse the slot in the next insertion
+            m_freeIndex = index;
```
Comment (elfenpiff) [inline iceoryx_posh/source/roudi/service_registry.cpp:83]: When you would use the fifo approach you could write here `m_freeIndices.push(index);`.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 10]
```diff
@@ -17,87 +17,83 @@
...
         }
-    };
+    }
+}
 
-    if (removedElement)
+void ServiceRegistry::removeAll(const capro::ServiceDescription& serviceDescription) noexcept
```
Comment (elfenpiff) [inline iceoryx_posh/source/roudi/service_registry.cpp:88]: I do not completely understand the use case for this method but `removeAll` is somehow misleading. What do you think about `purge`. Or that we rename `remove` into `deregister` and this into `remove`? If this is part of a follow up PR I would leave it as it is as long it is mentioned in some issue.

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp [hunk 11]
```diff
@@ -72,11 +88,61 @@ class ServiceRegistry
...
+    // store the last known free Index (if any is known)
+    // we should not use a queue (or stack) here since they are not optimal
+    // for the filling pattern of a vector (prefer entries close to the front)
+    uint32_t m_freeIndex{NO_INDEX};
+
+  private:
+    // functions for the different search cases
+    // tag types are not needed in all cases to distinguish overloads but kept for consistency
+
+    uint32_t find(const capro::ServiceDescription& serviceDescription) const noexcept;
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:112]: There are 10 overloads of `find` and it seems you can create three groups of them when you consider the behavior. This is hard. Could we please change this.

The user api `find` can be named as it is right now. 

This `find` is actually something like `getIndexOfServiceDescription`.

And the remaining finds, why not make one method out of it and name it `fillSearchResultVector`.
```cpp
void fillSearchResultVector(const cxx::optional<capro::IdString_t> & service, const cxx::optional<capro::IdString_t> & instance, const cxx::optional<capro::IdString_t> & event, ServiceDescriptionVector_t& searchResult);
```

In the end I think the code does not become much more complex since one could write it like:
```cpp
for(auto & entry : m_serviceDescriptions) {
  bool isEqual = static_cast<bool>(entry);
  isEqual &= (service) ? (entry->serviceDescription.getServiceIDString() == *service) : true;
  isEqual &= (instance) ? (entry->serviceDescription.getInstanceIDString() == *instance) : true;
  isEqual &= (event) ? (entry->serviceDescription.getEventIDString() == *event) : true;
  if ( isEqual ) {
    searchResult.emplace_back(*entry);
  }
}
```

This would be much less code and complexity as with this approach.

**Review by elfenpiff — CHANGES_REQUESTED** [review]
Please refactor the `find` method variations as suggested.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 12]
```diff
@@ -106,84 +102,178 @@ void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
...
+        return find(WILDCARD, WILDCARD, *event, searchResult);
+    }
+
+    return findAll(searchResult);
+}
+
+const ServiceRegistry::ServiceDescriptionVector_t ServiceRegistry::getServices() const noexcept
+{
+    ServiceDescriptionVector_t allEntries;
+    findAll(allEntries);
```
Comment (mossmaurice) [inline iceoryx_posh/source/roudi/service_registry.cpp:145]: ~~Why is `findAll()` needed? You could also invoke the copy c'tor of `cxx::vector`.~~

Forgot that you are storing `cxx::optional`'s... maybe `getAll()` is better, because we're not really searching anything here.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 13]
```diff
@@ -445,4 +465,109 @@ TEST_F(ServiceRegistry_test, AddingVariousServiceDescriptionAndGetServicesDoesNo
...
+    // deliberately contains no `0` (need to exclude some char)
+    static const char chars[] = "123456789"
+                                "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                "abcdefghijklmnopqrstuvwxyz";
+
+    constexpr auto N = string_t::capacity();
+    size = std::min(N, size);
+
+    char a[N + 1];
+    for (uint64_t i = 0; i < size; ++i)
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:490]: Nitpick ;)

```suggestion
    for (uint64_t i = 0U; i < size; ++i)
```

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp [hunk 14]
```diff
@@ -445,4 +465,109 @@ TEST_F(ServiceRegistry_test, AddingVariousServiceDescriptionAndGetServicesDoesNo
...
+    a[size] = '\0';
+
+    string_t s(a);
+    return s;
+}
+
+TEST_F(ServiceRegistry_test, CanAddMaximumNumberOfDifferentServiceDescriptions)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "76aef6cb-7886-4d64-9188-09bd1be2d335");
+    uint32_t numEntriesAdded = 0;
```
Comment (mossmaurice) [inline iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:503]: Nitpick ;)

```suggestion
    uint32_t numEntriesAdded = 0U;
```

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 5 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:45]: Not really important, but I do not care either way. The ctor was just needed for an emplace.

**Comment by MatthiasKillat** [comment]
@elfenpiff Note that it builds on MacOS and Windows with small size (like 100 entries, did not test the limit), but fails with 5000. I suspect it is the stack size but did not look into it yet. Not sure what is set in these environments or whether there is another problem.

We need to increase the limits and ensure it builds on these as well.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 5 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:45]: Only this implementation should be affected and will be moved.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 6 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:39]: @elfenpiff I considered this, actually my `TypedAllocator` works like this. But here I opted against it, because I want that earlier slots in the vector are filled first.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 7 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:57]: Will see whether it works.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 8 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:97]: Can do, we already have this dependency.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 9 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:83]: Sure, but I will instead remove the optimization.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 10 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:88]: I was thinking about the naming myself.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 11 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp:112]: For the the first part which returns an index I agree but for the second part I do not.

I will not check `isEqual &= (service)` every time I look at an entry and it is in no way clear the compiler can and will optimize it (I tend to think it will not). That is why I check it **once** and then invoke the specific case handler (i.e. overload of find). This is exactly what overloading is for. Furthermore it is event conceivable to have these overloads public for cases where we already know that a specific wildcard is needed (at compile time). This would require testing them in a public interface which I do not want at this point though.

Furthermore, I saw in more sophisticated approaches (with a search tree) that it might be required to have very different case handling. Having different overloads for different cases makes this way easier.

TLDR: The code you present may look more compact but at the cost of being slower by doing unnecessary work and making it harder to change individual cases. Rejected for this reason.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 15]
```diff
@@ -17,87 +17,83 @@
...
+    }
+
+    // search from start
+    for (auto& entry : m_serviceDescriptions)
+    {
+        if (!entry)
         {
-            // Due to n:m communication we don't store twice but increase the reference counter
-            element.referenceCounter++;
+            entry.emplace(serviceDescription, 1U);
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:52]: Can do.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 16]
```diff
@@ -17,87 +17,83 @@
...
+{
+    auto index = find(serviceDescription);
+    if (index != NO_INDEX)
     {
-        removeIndexFromMap(m_serviceMap, index);
-        removeIndexFromMap(m_instanceMap, index);
-        removeIndexFromMap(m_eventMap, index);
+        auto& entry = m_serviceDescriptions[index];
+        entry.reset();
+        // reuse the slot in the next insertion
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:95]: Not really a good optimization. A queue or stack would be more general but not quite helpful to fill the first slot.

File: iceoryx_posh/source/roudi/service_registry.cpp (hunk 12 above)
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:145]: Initially it was another `find` overload but I wanted to be more expressive. It corresponds to all wildcard search.

File: iceoryx_posh/source/roudi/service_registry.cpp [hunk 17]
```diff
@@ -106,84 +102,178 @@ void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const cxx::optional<capro::IdString_t>& instance,
                            const cxx::optional<capro::IdString_t>& event) const noexcept
 {
-    ServiceDescriptionVector_t serviceInstanceResult;
-    // Find (K1, K2)
-    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
-    if (instance && service)
+    constexpr Wildcard WILDCARD;
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/source/roudi/service_registry.cpp:105]: No it is not, only the type is there but no value of this type. I defined a local constant in the cpp though.

File: iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp (hunk 13 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_service_registry.cpp:490]: Yes but this really has to stop because it is clutter and serves no purpose (here). There is a time and place where unsigned and signed literals make a difference, but this is not one.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 11 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp]: I disagree, lets ask @elBoberido as referee.

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 18]
```diff
@@ -131,7 +131,7 @@ constexpr uint32_t MAX_PROCESS_NUMBER = 300U;
 /// This limitation is coming due to the fixed capacity of the cxx::vector (This doesn't limit the offered number of
 /// instances)
 /// @todo #415 increase number back to 50 once service registry is available via shared memory
-constexpr uint32_t MAX_NUMBER_OF_SERVICES = 10U;
+constexpr uint32_t MAX_NUMBER_OF_SERVICES = 100U;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:134]: I think you can remove the `@todo` above?

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 11 above)
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp]: @MatthiasKillat @elfenpiff As far as I know, this is more or less a stop gap solution right? When we notice that the search is too slow we would use the `PrefixTree` anyway? Is this assumption correct?
Furthermore, even if there is not yet a consensus on #1052, my bet would be that we will go with that route.

With all of this in mind, this looks like premature optimization and might even be removed in the not too distant future. While I understand the pursuit for performance, there are far more (yes, plural) elephants in the room which could give us a far bigger performance boost than this micro optimizations. Cases like false sharing, data distributed over two cache lines, context switches, etc. So all the stuff where the CPU needs to wait because of data synchronization or data not available. Stuff like that is more than an order of magnitude more relevant for performance issues than the stuff we constantly argue about. Please take a glass of wine and carefully read this blog post https://pkolaczk.github.io/server-slower-than-a-laptop/. This is the stuff we need to worry about.

So I'm leaning towards @elfenpiff since it tries to optimize something which does not even cause problems but this optimization adds costs in the form of maintenance and might make refactoring even harder.

Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp]: @elBoberido I am not sure it really will be a stop gap solution. And yes, we should strive to use atomics correctly as well etc. The #1052 would make the tree more attractive as well. Or a variation of it. But the tree is necessarily more complicated.

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 7 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp]: There are a lot of problems related to this, hence a new issue #1074.

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 18 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp]: Quite the contrary, more todos await in #1074

File: iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp (hunk 11 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp]: @elfenpiff Changed under the condition that we (strongly) consider using the prefix tree for this and similar problems in the future. I dislike to use suboptimal (runtime etc.) algorithms just because they are simple. This is often deceiving and minor inefficiencies add up.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elfenpiff — DISMISSED** [review]

**Review by mossmaurice — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
